### PR TITLE
HADOOP-16615. Add password check for credential provider

### PR DESCRIPTION
**Pull request by hddong**
When we use hadoop credential provider to store password, we can not sure if the password is the same as what we remembered.

So, I think we need a check tool.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 663 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1215 | trunk passed |
| -1 | compile | 190 | root in trunk failed. |
| +1 | checkstyle | 37 | trunk passed |
| +1 | mvnsite | 70 | trunk passed |
| +1 | shadedclient | 896 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 77 | trunk passed |
| 0 | spotbugs | 120 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 117 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 50 | the patch passed |
| -1 | compile | 180 | root in the patch failed. |
| -1 | javac | 180 | root in the patch failed. |
| -0 | checkstyle | 33 | hadoop-common-project/hadoop-common: The patch generated 11 new + 22 unchanged - 0 fixed = 33 total (was 22) |
| +1 | mvnsite | 65 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 824 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 77 | the patch passed |
| +1 | findbugs | 119 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 524 | hadoop-common in the patch failed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 5202 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.fs.viewfs.TestViewFsTrash |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1614 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 6f980e55ced7 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4fdf016 |
| Default Java | 1.8.0_222 |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/artifact/out/branch-compile-root.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/artifact/out/patch-compile-root.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/artifact/out/patch-compile-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/testReport/ |
| Max. process+thread count | 1461 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 1036 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1234 | trunk passed |
| -1 | compile | 205 | root in trunk failed. |
| +1 | checkstyle | 36 | trunk passed |
| +1 | mvnsite | 78 | trunk passed |
| +1 | shadedclient | 963 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 96 | trunk passed |
| 0 | spotbugs | 134 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 132 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 54 | the patch passed |
| -1 | compile | 210 | root in the patch failed. |
| -1 | javac | 210 | root in the patch failed. |
| -0 | checkstyle | 34 | hadoop-common-project/hadoop-common: The patch generated 11 new + 23 unchanged - 0 fixed = 34 total (was 23) |
| +1 | mvnsite | 70 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 851 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 86 | the patch passed |
| +1 | findbugs | 138 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 552 | hadoop-common in the patch failed. |
| +1 | asflicense | 35 | The patch does not generate ASF License warnings. |
| | | 5854 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.util.TestReadWriteDiskValidator |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1614 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b7b849b49a07 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4fdf016 |
| Default Java | 1.8.0_222 |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/artifact/out/branch-compile-root.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/artifact/out/patch-compile-root.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/artifact/out/patch-compile-root.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/testReport/ |
| Max. process+thread count | 1368 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hddong** [comment]
@steveloughran HADOOP-16615 patch here.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java [hunk 1]
```diff
@@ -86,6 +88,11 @@ protected int init(String[] args) throws IOException {
           return 1;
         }
         setSubCommand(new CreateCommand(args[++i]));
+      } else if (args[i].equals("check")) {
+        if (i == args.length -1) {
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java]: nit: add a space between - and 1

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java [hunk 2]
```diff
@@ -66,6 +67,7 @@
    * <pre>
    * % hadoop credential create alias [-provider providerPath]
    * % hadoop credential list [-provider providerPath]
+   * % hadoop credential check alias [-provider providerPath]
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java]: should the full usage option set be listed here?

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/alias/TestCredShell.java [hunk 3]
```diff
@@ -174,11 +174,20 @@ public void testPromptForCredential() throws Exception {
...
-        jceksProvider};
+
+    String[] args2 = {"check", "credential1", "-provider",
+      jceksProvider};
+    ArrayList<String> password = new ArrayList<String>();
+    password.add("p@ssw0rd");
+    shell.setPasswordReader(new MockPasswordReader(password));
     rc = shell.run(args2);
     assertEquals(0, rc);
+    assertTrue(outContent.toString().contains("Password match success for credential1."));
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/alias/TestCredShell.java]: Add an error message here which includes the failing string if the conditions not met. FWIW we are moving to using AssertJ assertions -if you are happy with them, feel free to use them over classic junit asserts

**Comment by steveloughran** [comment]
oops, deleted the wrong yetus comment. once you update the patch it will kick off again.

Overall -patch LGTM; a little bit of tuning and it's ready

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java (hunk 2 above)
Comment (hddong) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java]: > should the full usage option set be listed here?

Just to keep the format uniform and `[-value]` is for test, may not need here.

File: hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/alias/TestCredShell.java (hunk 3 above)
Comment (hddong) [reply hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/alias/TestCredShell.java]: An error check added.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 1794 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1067 | trunk passed |
| +1 | compile | 1016 | trunk passed |
| +1 | checkstyle | 51 | trunk passed |
| +1 | mvnsite | 86 | trunk passed |
| +1 | shadedclient | 881 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 92 | trunk passed |
| 0 | spotbugs | 123 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 120 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 49 | the patch passed |
| +1 | compile | 968 | the patch passed |
| +1 | javac | 968 | the patch passed |
| -0 | checkstyle | 51 | hadoop-common-project/hadoop-common: The patch generated 12 new + 22 unchanged - 0 fixed = 34 total (was 22) |
| +1 | mvnsite | 80 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 718 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 94 | the patch passed |
| +1 | findbugs | 132 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 545 | hadoop-common in the patch passed. |
| +1 | asflicense | 55 | The patch does not generate ASF License warnings. |
| | | 7897 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1614 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 9f23c3ae07b4 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / f267917 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/3/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/3/testReport/ |
| Max. process+thread count | 1374 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 83 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1226 | trunk passed |
| +1 | compile | 1082 | trunk passed |
| +1 | checkstyle | 45 | trunk passed |
| +1 | mvnsite | 80 | trunk passed |
| +1 | shadedclient | 938 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 87 | trunk passed |
| 0 | spotbugs | 128 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 126 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 48 | the patch passed |
| +1 | compile | 1026 | the patch passed |
| +1 | javac | 1026 | the patch passed |
| -0 | checkstyle | 46 | hadoop-common-project/hadoop-common: The patch generated 12 new + 22 unchanged - 0 fixed = 34 total (was 22) |
| +1 | mvnsite | 78 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 806 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 85 | the patch passed |
| +1 | findbugs | 130 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 596 | hadoop-common in the patch passed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 6614 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1614 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b101a7cec088 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / f267917 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/4/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/4/testReport/ |
| Max. process+thread count | 1378 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
code is good, you just need to deal with those checkstyles, which are mostly indentation and a couple of minor line lengths

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 41 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 1 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1069 | trunk passed |
| +1 | compile | 1023 | trunk passed |
| +1 | checkstyle | 52 | trunk passed |
| +1 | mvnsite | 84 | trunk passed |
| +1 | shadedclient | 890 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 94 | trunk passed |
| 0 | spotbugs | 123 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 119 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 48 | the patch passed |
| +1 | compile | 971 | the patch passed |
| +1 | javac | 971 | the patch passed |
| -0 | checkstyle | 49 | hadoop-common-project/hadoop-common: The patch generated 1 new + 22 unchanged - 0 fixed = 23 total (was 22) |
| +1 | mvnsite | 80 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 732 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 94 | the patch passed |
| +1 | findbugs | 131 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 545 | hadoop-common in the patch passed. |
| +1 | asflicense | 53 | The patch does not generate ASF License warnings. |
| | | 6178 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1614 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux cf69b3d98101 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6e5cd52 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/6/artifact/out/diff-checkstyle-hadoop-common-project_hadoop-common.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/6/testReport/ |
| Max. process+thread count | 1370 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1614/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java [hunk 4]
```diff
@@ -293,6 +300,91 @@ public String getUsage() {
...
+        "The check subcommand check a password for the name\n" +
+        "specified as the <alias> argument within the provider indicated\n" +
+        "through the -provider argument. If -strict is supplied, fail\n" +
+        "immediately if the provider requires a password and none is given.\n" +
+        "If -value is provided, use that for the value of the credential\n" +
+        "instead of prompting the user.";
+
+    private String alias = null;
+
+    public CheckCommand(String alias) {
```
Comment (hddong) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/alias/CredentialShell.java]: @steveloughran a checkstyle error here `public CheckCommand(String alias) {:5: Redundant 'public' modifier. [RedundantModifier]` , but i think `public` is need here.

**Comment by hddong** [comment]
retest this, please

**Comment by hddong** [comment]
retest this please

**Comment by hddong** [comment]
@steveloughran please help to retest this.

**Comment by steveloughran** [comment]
I like all tests in our code to provide enough diagnostics on failure that we can work out what went wrong purely from the Jenkins logs.

AssertJ is really good here, which is why we are adopting it in new code.

For your tests, I'm going to propose a new assertion which can be used to validate the output of the new command. If the condition is not met,
The exception raised will include all of the output. 

```
private void assertOutputContains(String expected) {
  Assertions.assertThat(outContent.toString())
    .contains(expected));
}
```

Yetus probably isn't retesting the code, as it thinks the last change was a documentation only change. If you change the test case things may be different.

**Comment by steveloughran** [comment]
thanks. code LGTM. Now github is saying things won't merge. Can you try to rebase this to trunk (yes, at the cost of the discussion); and we can go from there. Provided yetus doesn't object, I plan to merge this ASAP

**Comment by steveloughran** [comment]
+1, merged to trunk. Thanks!

(if you do a PR with this cherry picked back to 3.2/3.1 then I'll pull back there too. It's low risk and beneficial to all)

**Comment by hddong** [comment]
@steveloughran thanks for your help and review.

**Pull request closed without merging**